### ISSUE TITLE
Nilearn resampling doesn't support int16, throws RuntimeError: data type not supported

### ISSUE THREAD

**Issue by dohmatob** [nilearn/nilearn#817]
Reproduce with:
```
import nibabel
import numpy as np
from nilearn.image import resample_img
affine = np.eye(4)
img = nibabel.Nifti1Image(np.zeros((2, 2, 2)).astype(np.int16), affine)
resample_img(img, target_affine=2. * affine)
```

I think similar data types are also concerned

Possible fix: The data should be cast before passing to `ndimage.interpolate`, etc.

**Comment by GaelVaroquaux** [comment]
> Possible fix: The data should be cast before passing to ndimage.interpolate,
> etc.

Agreed for the fix, with a warning about the casting.

Do you want to submit a PR?

**Comment by dohmatob** [comment]
OK, sure. I'll send one this evening.

**Comment by dohmatob** [comment]
The following code segment is really dangerous (lines 457 --- of `nilearn/image/resample.py`):

```
    if interpolation == 'continuous' and data.dtype.kind == 'i':
        resampled_data_dtype = np.dtype(
            data.dtype.name.replace('int', 'float'))
    else:
        resampled_data_dtype = data.dtype

```

For example, image data.type were "int8". This would lead to "float8", which doesn't exist.

**Comment by dohmatob** [comment]
Affected types:

```
elvis@middle-earth:~/CODE/FORKED/nilearn$ python fuzz.py 

Fuzzing dtype: <type 'int'>
Ok.

Fuzzing dtype: <type 'numpy.int8'>
data type "float8" not understood

Fuzzing dtype: <type 'numpy.int16'>
data type not supported

Fuzzing dtype: <type 'numpy.int32'>
Ok.

Fuzzing dtype: <type 'numpy.int64'>
Ok.

Fuzzing dtype: <type 'numpy.uint32'>
Ok.

Fuzzing dtype: <type 'numpy.uint8'>
Ok.

Fuzzing dtype: <type 'numpy.uint16'>
Ok.

Fuzzing dtype: <type 'numpy.uint32'>
Ok.

Fuzzing dtype: <type 'numpy.uint64'>
Ok.

Fuzzing dtype: <type 'numpy.float32'>
Ok.

Fuzzing dtype: <type 'numpy.float64'>
Ok.

Fuzzing dtype: <type 'float'>
Ok.
```

**Comment by dohmatob** [comment]
Ok, in fact the previously posted code segment ( (lines 457 --- of `nilearn/image/resample.py`) actually introduces the bug. It fabricates `resampled_data_dtype = "float16"`, which doesn't exist.

**Comment by dohmatob** [comment]
See PR #818.

**Comment by GaelVaroquaux** [comment]
> Closed #817.

Good job Elvis! Thanks
